crates/pgt_typecheck/src/typed_identifier.rs

```diff
@@ -0,0 +1,340 @@
+use pgt_schema_cache::PostgresType;
+use pgt_treesitter_queries::{TreeSitterQueriesExecutor, queries::ParameterMatch};
+
+/// A typed identifier is a parameter that has a type associated with it.
+/// It is used to replace parameters within the SQL string.
+#[derive(Debug)]
+pub struct TypedIdentifier {
+    /// The path of the parameter, usually the name of the function.
+    /// This is because `fn_name.arg_name` is a valid reference within a SQL function.
+    pub path: String,
+    /// The name of the argument
+    pub name: Option<String>,
+    /// The type of the argument with schema and name
+    pub type_: IdentifierType,
+}
+
+#[derive(Debug, Clone)]
+pub struct IdentifierType {
+    pub schema: Option<String>,
+    pub name: String,
+    pub is_array: bool,
+}
+
+/// Applies the identifiers to the SQL string by replacing them with their default values.
+pub fn apply_identifiers<'a>(
+    identifiers: Vec<TypedIdentifier>,
+    schema_cache: &'a pgt_schema_cache::SchemaCache,
+    cst: &'a tree_sitter::Tree,
+    sql: &'a str,
+) -> (String, bool) {
+    let mut executor = TreeSitterQueriesExecutor::new(cst.root_node(), sql);
+
+    executor.add_query_results::<ParameterMatch>();
+
+    // Collect all replacements first to avoid modifying the string while iterating
+    let replacements: Vec<_> = executor
+        .get_iter(None)
+        .filter_map(|q| {
+            let m: &ParameterMatch = q.try_into().ok()?;
+            let path = m.get_path(sql);
+            let parts: Vec<_> = path.split('.').collect();
+
+            // Find the matching identifier and its position in the path
+            let (identifier, position) = find_matching_identifier(&parts, &identifiers)?;
+
+            // Resolve the type based on whether we're accessing a field of a composite type
+            let type_ = resolve_type(identifier, position, &parts, schema_cache)?;
+
+            Some((m.get_byte_range(), type_, identifier.type_.is_array))
+        })
+        .collect();
+
+    let mut result = sql.to_string();
+
+    let mut valid_positions = true;
```

Comment:
I'm not sure what `position` is in this context.
I understand it this way: We'll need to adjust the statements ranges if a parameter's default value is longer than the substituted text.
If that's correct, how about we rename this to something like `should_adjust_ranges` (default false) ?

Reply:
sometimes (almost never), we get a position back for the type error. in that case, we show the diagnostics on that position instead of the entire statement. since we are replacing the vars in the text, the position we get back does not match the position in the original statement anymore *if the replacements are not the same length*. hence, if `valid_positions` is `false`, we throw away the position from the type error and always show the error on the entire length of the statement.